app/controllers/UploadController.scala

```diff
@@ -46,41 +51,54 @@ class UploadController(val authActions: HMACAuthActions, awsConfig: AWSConfig, o
   }
 
   def credentials(id: String) = APIHMACAuthAction { implicit req =>
-    (table.get(id), req.headers.get(UPLOAD_KEY_HEADER)) match {
-      case (Some(upload), Some(key)) =>
-        val validKey = upload.parts.exists(_.key == key)
-
-        if(validKey) {
-          val credentials = creds.forKey(id, key)
-          Ok(Json.toJson(credentials))
-        } else {
-          BadRequest(s"Unknown part key $key")
-        }
-
-      case _ =>
-        BadRequest(s"Unknown upload or missing $UPLOAD_KEY_HEADER. id=$id")
+    partRequest(id, req) { (upload, part) =>
+      val credentials = uploads.credentialsForPart(upload, part)
+      Ok(Json.toJson(credentials))
     }
   }
 
-  private def chunk(uploadId: String, size: Long): List[UploadPart] = {
-    val boundaries = Upload.calculateChunks(size)
+  def complete(id: String) = APIHMACAuthAction { implicit req =>
+    partRequest(id, req) { (upload, part) =>
+      val complete = uploads.partComplete(upload, part)
+      table.put(complete)
 
-    boundaries.zipWithIndex.map { case ((start, end), id) =>
-      UploadPart(UploadPartKey(awsConfig.userUploadFolder, uploadId, id).toString, start, end)
+      NoContent
     }
   }
 
-  private def buildUpload(atom: MediaAtom, user: User, size: Long) = {
-    val id = UUID.randomUUID().toString
-
-    Upload(
-      id = id,
+  private def buildUpload(atom: MediaAtom, channelId: String, user: User, size: Long) = {
```

Comment:
I'll change it so we pattern match inside the method and throw an exception (caught in the `parse` wrapper method) if it isn't there